docs/advanced/inline-asm.md

```diff
@@ -2,58 +2,70 @@
+To achieve this currently, you have to edit the auto-generated `.scrypt` files under your project's `artifacts` directory.
 
-Inside this directory, there is a shell script named `apply_asm_optim.sh`. In this script, you can specify the source files where substitution with a custom bitcoin script should occur.
-
-Let's take a quick look at how this is applied in `scrypt-ts-lib` for three source files:
-
-```sh
-# BN256
-apply artifacts/src/ec/bn256.scrypt optimizations/ec/bn256
+First you create a project called `P2PKH`:
+```bash
+npx scrypt-cli project P2PKH --asm
+```
+Notice the `--asm` option must be enabled, meaning you are going to use inline assembly format of script.
+
+Your contract is at `src/contracts/p2pkh.ts`:
+
+```ts
+export class P2PKH extends SmartContract {
+    @prop()
+    readonly pubKeyHash: PubKeyHash
+
+    constructor(pubKeyHash: PubKeyHash) {
+        super(...arguments)
+        this.pubKeyHash = pubKeyHash
+    }
+
+    @method()
+    public unlock(sig: Sig, pubkey: PubKey) {
+        assert(
+            hash160(pubkey) == this.pubKeyHash,
+            'public key hashes are not equal'
+        )
+        assert(this.checkSig(sig, pubkey), 'signature check failed')
+    }
+}
+```
 
-# SECP256K1
-apply artifacts/src/ec/secp256k1.scrypt optimizations/ec/secp256k1
+Say you want to substitute the `unlock` function with manual script, you edit the file `.asm/asm.json`.
 
-# SECP256R1
-apply artifacts/src/ec/secp256r1.scrypt optimizations/ec/secp256r1
+```json
+{
+  "P2PKH": {
+    "unlock": "op_dup op_hash160 $pubKeyHash op_equalverify op_checksig"
+  }
+}
 ```
 
-For instance, let's consider the content of `optimizations/ec/bn256`:
+Variables can be defined by prefix `$`, as in `$pubKeyHash`.
 
-```
-_addCurvePoints.asm
-doubleCurvePoint.asm
-lineFuncAdd.asm
-lineFuncDouble.asm
-modInverseBranchlessP.asm
-mulFQ12.asm
-mulLine.asm
-squareFQ12.asm
-```
+We could also define multiple substitutions for multiple methods, if needed.
 
-The name of each file corresponds to a function in the specified `.scrypt` source file (`bn256.scrypt` for this example). The files contain the actual bitcoin script in ASM format:
+Now, after every project build, the function body will be replaced with script, as could be seen in `artifacts/P2PKH.scrypt`.
 
-```
-OP_3 OP_PICK 11 OP_PICK OP_MUL 12 OP_PICK OP_4 OP_PICK OP_MUL OP_ADD OP_3 OP_PICK 12 OP_PICK OP_MUL OP_5 OP_PICK 14 OP_PICK OP_MUL OP_SUB OP_7 OP_PICK 11 OP_PICK OP_MUL 12 OP_PICK OP_8 OP_PICK OP_MUL OP_ADD OP_7 OP_PICK 12 OP_PICK OP_MUL OP_9 OP_PICK 14 OP_PICK OP_MUL OP_SUB OP_3 OP_ROLL OP_ROT OP_ADD OP_ROT OP_ROT OP_ADD OP_4 OP_PICK 11 OP_PICK 13 OP_PICK OP_8 OP_PICK OP_DUP OP_3 OP_PICK OP_MUL OP_2 OP_PICK OP_5 OP_PICK OP_MUL OP_ADD OP_SWAP OP_ROT OP_MUL OP_3 OP_ROLL OP_3 ...
-```
+![](../../static/img/asm.png)
```

Comment:
Update to one without explicit constructor.

Reply:
@msinkec Do not forget to update this. Also rename
replaceAsmVars -> setAsmVars